trust-quorum/src/connection_manager.rs

```diff
@@ -0,0 +1,717 @@
+}
+
+/// Messages sent from connection managing tasks to the main peer task
+///
+/// We include `task_id` to differentiate which task they come from so we can
+/// exclude requests from tasks that have been cancelled or have been told to
+/// shutdown.
+#[derive(Debug, PartialEq)]
+pub struct ConnToMainMsg {
+    pub task_id: TaskId,
+    pub msg: ConnToMainMsgInner,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ConnToMainMsgInner {
+    Accepted { addr: SocketAddrV6, peer_id: BaseboardId },
+    Connected { addr: SocketAddrV6, peer_id: BaseboardId },
+    Received { from: BaseboardId, msg: PeerMsg },
+    ReceivedNetworkConfig { from: BaseboardId, config: NetworkConfig },
+    Disconnected { peer_id: BaseboardId },
+}
+
+pub struct TaskHandle {
+    pub task_id: TaskId,
+    pub tx: mpsc::Sender<MainToConnMsg>,
+    pub conn_type: ConnectionType,
+}
+
+impl TaskHandle {
+    pub fn addr(&self) -> SocketAddrV6 {
+        self.conn_type.addr()
+    }
+}
+
+pub enum ConnectionType {
+    Connected(SocketAddrV6),
+    Accepted(SocketAddrV6),
+}
+
+impl ConnectionType {
+    pub fn addr(&self) -> SocketAddrV6 {
+        match self {
+            Self::Connected(addr) => *addr,
+            Self::Accepted(addr) => *addr,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum ConnState {
+    Connecting,
+    Accepting,
+    Established(BaseboardId),
+}
+
+/// Information about a single connection task
+#[derive(Debug, Clone)]
+pub struct ConnInfo {
+    pub state: ConnState,
+    pub addr: SocketAddrV6,
+    pub task_id: TaskId,
+}
+
+/// Status information useful for debugging
+#[derive(Debug, Clone)]
+pub struct ConnMgrStatus {
+    pub bootstrap_addrs: BTreeSet<SocketAddrV6>,
+    pub connections: Vec<ConnInfo>,
+    pub num_task_join_handles: u64,
+    pub next_task_id: TaskId,
+}
+
+/// A structure to manage all sprockets connections to peer nodes
+///
+/// Each sprockets connection runs in its own task which communicates with the
+/// main `NodeTask`. All methods on the `ConnMgr` run inside the main `NodeTask`
+/// as `ConnMgr` is a member field of `NodeTask`. This allows isolating the
+/// connection management logic from the main node message handling logic
+/// without adding yet another task.
+pub struct ConnMgr {
+    log: Logger,
+
+    /// A channel for sending messages from a connection task to the main task
+    main_tx: mpsc::Sender<ConnToMainMsg>,
+
+    /// The sprockets config
+    config: SprocketsConfig,
+
+    /// The sprockets server
+    server: sprockets_tls::Server,
+
+    /// The address the sprockets server listens on
+    listen_addr: SocketAddrV6,
+
+    // A unique, monotonically incrementing id for each task to help map tasks
+    // to their handles in case the task aborts, or there is a new connection
+    // accepted and established for an existing `BaseboardId`.
+    next_task_id: TaskId,
+
+    /// `JoinHandle`s to all tasks that can be polled for crashes
+    join_handles: FuturesUnordered<JoinHandle<TaskId>>,
```

Comment:
nit, take it or leave it: i might consider using a [`JoinSet`](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html) for this...

Reply:
What's the benefit of a `JoinSet` here? The current solution works, and I'm lazy, so I likely won't change now. But I'd like to know for the future.

Reply:
I'll probably switch to `join_next_with_id` and just use `tokio::task::Ids` also, since I just check for equality and don't compare for order like I have done in the past. :)

Reply:
Yay, glad that ended up being useful1